lib/environ/environ.go

```diff
@@ -0,0 +1,258 @@
+	_, err := r.init()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (r *Updater) init() (*libfsm.FSM, error) {
+	_, err := r.getOrCreateOperationPlan()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	machine, err := fsm.New(fsm.Config{
+		Operation:    r.Operation,
+		Operator:     r.Operator,
+		Backend:      r.Backend,
+		LocalBackend: r.LocalBackend,
+		Runner:       r.Runner,
+		Silent:       r.Silent,
+		Emitter:      r.Emitter,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return machine, nil
+}
+
+func (r *Updater) executePlan(ctx context.Context, machine *libfsm.FSM, force bool) error {
+	planErr := machine.ExecutePlan(ctx, nil, force)
+	if planErr != nil {
+		r.Warnf("Failed to execute plan: %v.", trace.DebugReport(planErr))
+	}
+
+	err := machine.Complete(planErr)
+	if err == nil {
+		err = planErr
+	}
+
+	var addrs []string
+	for _, server := range r.Servers {
+		addrs = append(addrs, server.AdvertiseIP)
+	}
+
+	// Keep the agents running as long as the operation can be resumed
+	if planErr == nil {
+		if errShutdown := rpc.ShutdownAgents(ctx, addrs, r.FieldLogger, r.Runner); errShutdown != nil {
+			r.Warnf("Failed to shutdown agents: %v.", trace.DebugReport(errShutdown))
+		}
+	}
+	return trace.Wrap(err)
+}
+
+func (r *Config) checkAndSetDefaults() error {
+	if r.Operator == nil {
+		return trace.BadParameter("cluster operator service is required")
+	}
+	if len(r.Servers) == 0 {
+		return trace.BadParameter("at least a single server is required")
+	}
+	if r.Backend == nil {
+		return trace.BadParameter("primary backend is required")
+	}
+	if r.LocalBackend == nil {
+		return trace.BadParameter("local backend is required")
+	}
+	if r.FieldLogger == nil {
+		r.FieldLogger = log.WithField(trace.Component, "envars:updater")
+	}
+	if r.Emitter == nil {
+		r.Emitter = utils.NopEmitter()
+	}
+	return nil
+}
+
+// Config describes configuration for updating cluster environment variables
+type Config struct {
+	// ClusterKey identifies the cluster
+	ClusterKey ops.SiteKey
+	// Operator is the cluster operator service
+	Operator ops.Operator
+	// Operation references a potentially active garbage collection operation
+	Operation *ops.SiteOperation
+	// Backend specifies the primary backend
+	Backend storage.Backend
+	// LocalBackend specifies the backend where intermediate operation state
+	// is stored. It is the fallback backend, if the primary backend is temporarily
+	// unavailable
+	LocalBackend storage.Backend
+	// Servers is the list of cluster servers
+	Servers []storage.Server
+	// Runner specifies the runner for remote commands
+	Runner libfsm.AgentRepository
+	// FieldLogger is the logger to use
+	log.FieldLogger
+	// Silent controls whether the process outputs messages to stdout
+	localenv.Silent
+	// Emitter outputs progress messages to stdout
+	utils.Emitter
+}
+
+// Updater executes a cobntrolled update of cluster environment variables
```

Comment:
cobntrolled -> controlled

Reply:
Yes, it seems that I'd need to update the planet's configuration package to persist the environment update.
Also, besides the fact that we had had bad experiences with container restarts previously on older kernels which hopefully would be less and less of a problem, there's nothing preventing the operation from doing the rolling restart.
I've spent enough time procrastinating over a set of generalizations for the fsm procedure and factoring out the rolling container restart and making it reusable is definitely a good idea.
I would prefer to do the restart manually at this point though as I don't have enough time capacity for the refactoring work, but I will create a ticket for me to return to this problem some time later.